pkg/cmd/provisioning/identity_provider.go

```diff
@@ -57,145 +67,273 @@ var (
-	})
+	certs := conn.ConnectionState().PeerCertificates
+	numCerts := len(certs)
 
-	if err != nil {
-		return errors.Wrapf(err, "Failed to upload JSON web key set (JWKS) in the S3 bucket %s", bucketName)
+	fingerprint := sha1.Sum(certs[numCerts-1].Raw)
+	var buf bytes.Buffer
+	for _, f := range fingerprint {
+		fmt.Fprintf(&buf, "%02X", f)
 	}
-	log.Printf("JSON web key set (JWKS) in the S3 bucket %s at %s updated", bucketName, keysURI)
+	return buf.String(), nil
+}
 
-	err = saveToFile("JSON web key set (JWKS)", filepath.Join(targetDir, "keys.json"), jwks)
+func createIAMIdentityProvider(client aws.Client, issuerURL, namePrefix, targetDir string, generateOnly bool) error {
+	fingerprint, err := getTLSFingerprint(issuerURL)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to get fingerprint")
 	}
 
-	oidcProviderList, err := client.ListOpenIDConnectProviders(&iam.ListOpenIDConnectProvidersInput{})
+	if generateOnly {
+		oidcIdentityProviderJSON := fmt.Sprintf(iamIdentityProviderTemplate, issuerURL, fingerprint)
+
+		err := saveToFile("AWS IAM Identity Provider", filepath.Join(targetDir, iamIdentityProviderFilename), []byte(oidcIdentityProviderJSON))
+		if err != nil {
+			errors.Wrap(err, "failed to save IAM Identity Provider file")
+		}
+
+	} else {
+		oidcProviderList, err := client.ListOpenIDConnectProviders(&iam.ListOpenIDConnectProvidersInput{})
+		if err != nil {
+			return errors.Wrap(err, "Failed to fetch list of Identity Providers")
+		}
+
+		var providerARN string
+		for _, provider := range oidcProviderList.OpenIDConnectProviderList {
+			ok, err := isExistingIdentifyProvider(client, *provider.Arn, namePrefix)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to check existing Identity Provider %s", *provider.Arn)
+			}
+
+			if ok {
+				providerARN = *provider.Arn
+				log.Printf("Existing Identity Provider found with ARN: %s", providerARN)
+				break
+			}
+		}
+
+		if len(providerARN) == 0 {
+			oidcOutput, err := client.CreateOpenIDConnectProvider(&iam.CreateOpenIDConnectProviderInput{
+				ClientIDList: []*string{
+					awssdk.String("openshift"),
+					awssdk.String("sts.amazonaws.com"),
+				},
+				ThumbprintList: []*string{
+					awssdk.String(fingerprint),
+				},
+				Url: awssdk.String(issuerURL),
+			})
+			if err != nil {
+				return errors.Wrap(err, "Failed to create Identity Provider")
+			}
+
+			providerARN = *oidcOutput.OpenIDConnectProviderArn
+
+			_, err = client.TagOpenIDConnectProvider(&iam.TagOpenIDConnectProviderInput{
+				OpenIDConnectProviderArn: &providerARN,
+				Tags: []*iam.Tag{
+					{
+						Key:   awssdk.String(fmt.Sprintf("%s/%s", ccoctlAWSResourceTagKeyPrefix, namePrefix)),
+						Value: awssdk.String(ownedCcoctlAWSResourceTagValue),
+					},
+				},
+			})
+			if err != nil {
+				return errors.Wrapf(err, "Failed to tag the identity provider with arn: %s", providerARN)
+			}
+
+			log.Printf("Identity Provider created with ARN: %s", providerARN)
```

Comment:
I probably missed this in the other pr but is log.Printf the way we're intentionally printing output in ccoctl?

Reply:
It seems there is no log.Info in this log package (https://golang.org/pkg/log/), so I suppose Printf is what we're left with.